File: solutions/0238-product-of-array-except-self/solution.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> PrefixV(nums.size(),1);
        vector<int> SuffixV(nums.size(),1);

        if(nums.size()<2){
            return nums;
        }
        PrefixV[0] = nums[0];
        SuffixV[nums.size()-1]= nums[nums.size()-1];
        for(int i=1;i<nums.size();i++){
            PrefixV[i] = PrefixV[i-1]*nums[i];
        }

        for(int i=nums.size()-2;i>=0;i--){
            SuffixV[i] = SuffixV[i+1]*nums[i];
        }

        nums[0] = SuffixV[1];
        nums[nums.size()-1] = PrefixV[nums.size()-2];

        for(int i=1;i<nums.size()-1;i++){
            nums[i] = PrefixV[i-1]*SuffixV[i+1];
        }
        return nums;
    }
};

File: solutions/0238-product-of-array-except-self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        leftProduct = [0]*len(nums)
        rightProduct = [0]*len(nums)
    
        # left to right
        leftProduct[0] = nums[0]
        for i in range(1,len(nums)):
            leftProduct[i] = nums[i]*leftProduct[i-1]
        
        # print(leftProduct)
        # right to left
        rightProduct[-1] = nums[-1]
        for i in range(len(nums)-2,-1,-1):
            rightProduct[i] = nums[i]*rightProduct[i+1]

        # print(rightProduct)

        # right to left
        ans = [0 for i in range(len(nums))]
        ans[0] = rightProduct[1]
        ans[-1] = leftProduct[-2]

        for i in range(1,len(nums)-1):
            ans[i] = leftProduct[i-1]*rightProduct[i+1]

        return ans
        
